topology/probes/ovs_of.go

```diff
@@ -0,0 +1,531 @@
+	PrivateKey   string                    // Path of the private key authenticating the probe.
+	CA           string                    // Path of the certicate of the Certificate authority used for authenticated communication with bridges
+	sslOk        bool                      // cert private key and ca are provisionned.
+}
+
+// BridgeOfProbe is the type of the probe retrieving Openflow rules on a Bridge
+type BridgeOfProbe struct {
+	Host       string           // The global host
+	Bridge     string           // The bridge monitored
+	UUID       string           // The UUID of the bridge node
+	Address    string           // The address of the bridge if different from name
+	BridgeNode *graph.Node      // the bridge node on which the rule nodes are attached.
+	OvsOfProbe *OvsOfProbe      // Back pointer to the probe
+	Rules      map[string]*Rule // The set of rules found so far.
+	cancel     context.CancelFunc
+}
+
+// Rule is an OpenFlow rule in a switch
+type Rule struct {
+	Cookie   uint64 // cookie value of the rule
+	Table    int    // table containing the rule
+	Priority int    // priority of rule
+	Filter   string // all the filters as a comma separated string
+	Actions  string // all the actions (comma separated)
+	InPort   int    // -1 is any
+	UUID     string // Unique id
+}
+
+// Event is an event as monitored by ovs-ofctl monitor <br> watch:
+type Event struct {
+	Rule   *Rule  // The rule modified
+	Date   int64  // the date of the event
+	Action string // the action taken
+	Bridge string // The bridge whtere it ocured
+}
+
+// ProtectCommas substitute commas with semicolon
+// when inside parenthesis
+func protectCommas(line string) string {
+	work := []rune(line)
+	var braces = 0
+	for i, word := range work {
+		switch word {
+		case '(':
+			braces++
+		case ')':
+			braces--
+		case ',':
+			if braces > 0 {
+				work[i] = ';'
+			}
+		}
+	}
+	return string(work)
+}
+
+// fillIn is a utility function that takes a splitted rule line
+// and fills a Rule/Event structure with it
+func fillIn(components []string, rule *Rule, event *Event) {
+	for _, component := range components {
+		keyvalue := strings.SplitN(component, "=", 2)
+		if len(keyvalue) == 2 {
+			key := keyvalue[0]
+			value := keyvalue[1]
+			switch key {
+			case "event":
+				if event != nil {
+					event.Action = value
+				}
+			case "actions":
+				rule.Actions = value
+			case "table":
+				table, err := strconv.ParseInt(value, 10, 32)
+				if err == nil {
+					rule.Table = int(table)
+				} else {
+					logging.GetLogger().Errorf("Error while parsing table of rule: %s", err.Error())
+				}
+			case "cookie":
+				v, err := strconv.ParseUint(value, 0, 64)
+				if err == nil {
+					rule.Cookie = v
+				} else {
+					logging.GetLogger().Errorf("Error while parsing cookie of rule: %s", err.Error())
+				}
+			}
+		}
+	}
+
+}
+
+// extractPriority parses the filter of a rule and extracts the priority if it exists.
+func extractPriority(rule *Rule) {
```

Comment:
I have pushed a fix with a small change in nodeTitle so that it does not fail when Name is undefined in graph-layout.js. The front-end will require a change to filter out rule nodes so that they are not displayed in the d3 view. So far I have used an explicit filter on the type (must not be 'ofrule') but I could also filter on the absence of Name which could be more extensible in the future. I have rebased to have an up-to-date front-end and it seems fine.